pkg/ring/kv/kv_test.go

```diff
@@ -165,23 +178,35 @@ func TestWatchPrefix(t *testing.T) {
 			}
 		}
 
-		ch1 := make(chan struct{})
-		ch2 := make(chan struct{})
-		go gen(prefix, ch1)
-		go gen(prefix2, ch2) // we don't want to see these keys reported
+		go gen(prefix)
+		go gen(prefix2) // we don't want to see these keys reported
 
 		observedKeys := map[string]int{}
-		ctx, cfn := context.WithTimeout(context.Background(), 1.5*max*sleep)
-		defer cfn()
-
-		client.WatchPrefix(ctx, prefix, func(key string, val interface{}) bool {
-			observedKeys[key] = observedKeys[key] + 1
-			return true
-		})
 
-		// wait until updaters finish (should be done by now)
-		<-ch1
-		<-ch2
+		totalDeadline := time.After(totalTestTimeout)
+		var nextTimer = time.NewTimer(nextKeyDeadline)
```

Comment:
To simplify it, what if we remove the `nextTimer` at all, and we simply set `watching = false` when `len(observedKeys)` reach `max`?

Reply:
Sounds good